clangd-wrapper/main.cpp

```diff
@@ -0,0 +1,127 @@
+#include "clangd-wrapper/version.h"
+#include <algorithm>
+#include <filesystem>
+#include <iostream>
+#include <optional>
+#include <string>
+#include <vector>
+#include <boost/process.hpp>
+#include <boost/program_options.hpp>
+
+namespace fs = std::filesystem;
+namespace bp = boost::process;
+namespace po = boost::program_options;
+
+bp::filesystem::path
+get_absolute_executable_path(bp::filesystem::path const& path)
+{
+  if (path.is_absolute()) {
+    return path;
+  } else {
+    return bp::search_path(path);
+  }
+}
+
+std::optional<std::string>
+get_bazel_execution_root(std::string const& bazel_path, std::string const& workspace_dir)
+{
+  bp::ipstream is;
+  bp::child c(
+    get_absolute_executable_path(bazel_path), "info", "execution_root", bp::std_out > is, bp::start_dir(workspace_dir));
+  c.wait();
+
+  std::string result;
+  std::string line;
+  while (std::getline(is, line)) {
+    result += line;
+  }
+
+  int exit_code = c.exit_code();
+  if (exit_code != 0) {
+    std::cerr << "Failed to run 'bazel info execution_root'. Exit code: " << exit_code << std::endl;
+    return std::nullopt;
+  }
+
+  // Remove trailing newline
+  if (!result.empty() && result.back() == '\n') {
+    result.pop_back();
+  }
+
+  return result;
+}
+
+fs::path
+get_workspace_dir(std::vector<std::string> const& args)
+{
+  for (std::string const& arg : args) {
+    if (arg.rfind("--compile-commands-dir=", 0) == 0) {
+      return arg.substr(strlen("--compile-commands-dir="));
+    }
+  }
+  return fs::current_path();
+}
+
+int
+main(int argc, char* argv[])
+{
+  po::options_description desc("Usage: bazel-clangd-wrapper OPTIONS [ -- CLANGD_OPTIONS ]");
+
+  // clang-format off
+  desc.add_options()
+    ("help,h", "Produce help message")
+    ("version,V", "print version")
+    ("bazel-path", po::value<std::string>()->default_value("bazel"), "Path to the bazel executable")
+    ("clangd-path", po::value<std::string>()->default_value("clangd"), "Path to the clangd executable");
+  // clang-format on
+
+  // Arguments before -- are passed to the wrapper, arguments after -- are passed to clangd.
+  char** split_iter = std::find(argv, argv + argc, std::string("--"));
+  std::vector<char*> wrapper_args(argv, split_iter);
+  std::vector<std::string> clangd_args;
+  if (split_iter != argv + argc) {
+    clangd_args.assign(split_iter + 1, argv + argc);
+  }
```

Comment:
wouldn't it be easier to use positional options where clangd options are an unlimited number of positional options, see https://www.boost.org/doc/libs/latest/doc/html/program_options/overview.html#id-1.3.29.5.7.10 for documentation.

Reply:
Also, it would cause a typo in a wrapper option to get passed to clangd instead regardless of position